### DISCUSSION TITLE
Makes types in the System.Numerics namespace serializable

### DISCUSSION THREAD

**Discussion by zdirtywork** [dotnet/csharplang#7747] [Language Ideas] [answered]
Make types in the System.Numerics namespace serializable.

Currently, `Vector2/3/4`, `Quaternion`, etc. do not have the `SerializableAttribute`.

In order to serialize these data, we have to redefine their equivalent types and frequently convert objects between the two types.

**Comment by TahirAhmadov** [comment, marked as answer]
I think this should be requested at https://github.com/dotnet/runtime/issues/

**Comment by koszeggy** [comment]
The runtime stopped making new types serializable because of the obsoletion of `BinaryFormatter`.

As an alternative you can use 3rd party libraries. For example, [this serializer](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_Serialization_Binary_BinarySerializationFormatter.htm) from [this](https://github.com/koszeggy/KGySoft.CoreLibraries#binary-serialization) library (disclaimer: written by me) supports the mentioned types from the `System.Numerics` namespace natively [[NuGet](https://www.nuget.org/packages/KGySoft.CoreLibraries)].

**Reply by tannergooding** [reply]
Noting that most forms of binary serialization, especially if they are modeled similarly to `BinaryFormatter`, need to be very carefully considered with the same kind of warning/contexts given in: https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide

> The [BinaryFormatter](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter) type is dangerous and is not recommended for data processing. Applications should stop using BinaryFormatter as soon as possible, even if they believe the data they're processing to be trustworthy. BinaryFormatter is insecure and can't be made secure.

Serializers like `XmlSerializer` and `System.Text.Json` are the preferred approach and can already handle these types. `BinaryReader/Writer` can be an alternative and can be used to serialize the individual `float` fields of these types, but it still needs to be handled with care.

**Reply by koszeggy** [reply]
@tannergooding: as you posted this as a reply to my comment I feel myself addressed. Though in my comment I didn't elaborate what "supports natively" means (as opposed to `IFormatter`-like recursive serialization with polymorphism and auto type resolving) if you follow the posted [link](https://docs.kgysoft.net/corelibraries/html/T_KGySoft_Serialization_Binary_BinarySerializationFormatter.htm) you should immediately see the huge red **Security Note** box about how the serializer addresses the issues that `BinaryFormatter` [suffers from](https://stackoverflow.com/a/67107584/5114784).

> Noting that most forms of binary serialization [...] need to be very carefully considered. [...] Serializers like `XmlSerializer` and `System.Text.Json` are the preferred approach.

Though I'm not contradicting what you are saying, still, it can be misinterpreted easily, supporting the common (mis)understanding that in general binary serialization is dangerous whereas JSON/XML serialization is safe.

Which is simply [not true](https://stackoverflow.com/a/76955010/5114784). For example ProtoBuf (binary) is safe because it's not polymorphic and supports only a small set of predefined types, whereas JSON.NET (textual) supports polymorphism and it can be a risk if you allow resolving the `$type` elements automatically (again, you didn't say it otherwise, I'm just completing your comment).
